hazelcast/include/hazelcast/client/exception/IException.h

```diff
@@ -48,36 +54,76 @@ namespace hazelcast {
              */
             class HAZELCAST_API IException : public std::exception {
             public:
-                /**
-                 * Constructor
-                 */
                 IException();
 
-                /**
-                 * Constructor
-                 */
                 IException(const std::string &source, const std::string &message);
 
-                /**
-                 * Destructor
-                 */
+                IException(const std::string &source, const std::string &message,
+                           const boost::shared_ptr<IException> &cause);
+
                 virtual ~IException() throw();
 
                 /**
-                 * return exception explanation string.
+                 *
+                 * return  pointer to the explanation string.
                  */
                 virtual char const *what() const throw();
 
                 const std::string &getSource() const;
 
                 const std::string &getMessage() const;
 
-                virtual void raise();
-            private:
+                virtual void raise() const;
+
+                /**
+                 * We need this method to clone the specific derived exception when needed. The exception has to be the
+                 * derived type so that the exception rethrowing works as expected by throwing the derived exception.
+                 * Exception throwing internals works by making a temporary copy of the exception.
+                 * @return The copy of this exception
+                 */
+                virtual std::auto_ptr<IException> clone() const;
+
+                const boost::shared_ptr<IException> &getCause() const;
+
+                friend std::ostream HAZELCAST_API &operator<<(std::ostream &os, const IException &exception);
```

Comment:
After introducing `ExceptionBuilder`, It looks like we should not need this method anymore, am I right?

Reply:
This the stream operator used when you want to print the exception, which is equivalent of toString method at Java. I think we may still need it. The operator ExceptionBuilder uses is serving another purpose to compose the exception message.